File: pinning-webui/tests/serviceAuth.test.ts
import { describe, it, expect } from 'vitest';
import crypto from 'crypto';
import { verifyServiceAuth } from '../server/serviceAuth.js';

// Mirrors the Fula S3 gateway (Rust) minter — the cross-language reference.
function mint(userId: string, exp: number, secret: string): string {
  const uidB64 = Buffer.from(userId).toString('base64url');
  const msg = `v1.${uidB64}.${exp}`;
  const sig = crypto.createHmac('sha256', secret).update(msg).digest('base64url');
  return `${msg}.${sig}`;
}

// SHARED VECTOR — identical inputs/header to the Go (service_auth_test.go) and
// Rust (service_auth.rs) tests. This is the cross-language lock.
const VEC_SECRET = 'fula-pin-svc-shared-test-secret-rotate-me';
const VEC_USER = '2d2dfffdad62ff927abba1295c73a4eab7666813280ea8b356da845e440c41ff';
const VEC_EXP = 4102444800;
const VEC_HEADER =
  'v1.MmQyZGZmZmRhZDYyZmY5MjdhYmJhMTI5NWM3M2E0ZWFiNzY2NjgxMzI4MGVhOGIzNTZkYTg0NWU0NDBjNDFmZg.4102444800.7kiFsUP9DMnoeP00uwbUwy4tJmTC6rpy9Dce_e-jw3U';

describe('serviceAuth', () => {
  it('matches + verifies the shared cross-language vector (locks Go/Rust/TS)', () => {
    expect(mint(VEC_USER, VEC_EXP, VEC_SECRET)).toBe(VEC_HEADER); // minter parity
    expect(verifyServiceAuth(VEC_HEADER, VEC_SECRET)).toBe(VEC_USER); // verifier parity
  });

  it('rejects wrong secret', () => {
    expect(verifyServiceAuth(VEC_HEADER, 'a-totally-different-secret')).toBeNull();
  });

  it('rejects empty / disabled secret', () => {
    expect(verifyServiceAuth(VEC_HEADER, '')).toBeNull();
    expect(verifyServiceAuth(VEC_HEADER, 'disabled')).toBeNull();
  });

  it('rejects a tampered signature', () => {
    const p = VEC_HEADER.split('.');
    p[3] = Buffer.from('not-the-real-hmac-32-bytes-xxxxx').toString('base64url');
    expect(verifyServiceAuth(p.join('.'), VEC_SECRET)).toBeNull();
  });

  it('rejects a user-id swap that keeps the signature (binding holds)', () => {
    const p = VEC_HEADER.split('.');
    p[1] = Buffer.from('victim-other-user').toString('base64url');
    expect(verifyServiceAuth(p.join('.'), VEC_SECRET)).toBeNull();
  });

  it('rejects expired', () => {
    const past = Math.floor(Date.now() / 1000) - 10;
    expect(verifyServiceAuth(mint(VEC_USER, past, VEC_SECRET), VEC_SECRET)).toBeNull();
  });

  it('rejects malformed / bad version / absent', () => {
    expect(verifyServiceAuth('v1.abc.123', VEC_SECRET)).toBeNull();
    expect(verifyServiceAuth('v2.' + VEC_HEADER.split('.').slice(1).join('.'), VEC_SECRET)).toBeNull();
    expect(verifyServiceAuth(undefined, VEC_SECRET)).toBeNull();
  });

  it('round-trips a freshly minted token', () => {
    const exp = Math.floor(Date.now() / 1000) + 60;
    expect(verifyServiceAuth(mint('user-xyz', exp, VEC_SECRET), VEC_SECRET)).toBe('user-xyz');
  });
});

File: openapi/go/service_auth_test.go
package openapi

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"strconv"
	"strings"
	"testing"
	"time"
)

// mintServiceAuth builds an X-Fula-Service-Auth value the way the Fula S3 gateway
// (Rust) does. It is the REFERENCE for the cross-language shared vector below;
// production never mints here (the pinning service only verifies).
func mintServiceAuth(userID string, exp int64, secret string) string {
	uidB64 := base64.RawURLEncoding.EncodeToString([]byte(userID))
	expStr := strconv.FormatInt(exp, 10)
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte("v1." + uidB64 + "." + expStr))
	sigB64 := base64.RawURLEncoding.EncodeToString(mac.Sum(nil))
	return "v1." + uidB64 + "." + expStr + "." + sigB64
}

// SHARED VECTOR — these exact inputs must produce the identical header in the
// Rust gateway minter (and verify identically anywhere else). Far-future exp so
// it never expires under test.
const (
	vecSecret = "fula-pin-svc-shared-test-secret-rotate-me"
	vecUserID = "2d2dfffdad62ff927abba1295c73a4eab7666813280ea8b356da845e440c41ff"
	vecExp    = int64(4102444800) // 2100-01-01T00:00:00Z
)

func TestServiceAuthSharedVector(t *testing.T) {
	header := mintServiceAuth(vecUserID, vecExp, vecSecret)
	t.Logf("SHARED VECTOR (copy into the Rust/TS tests):\n  secret=%q\n  user_id=%q\n  exp=%d\n  header=%s",
		vecSecret, vecUserID, vecExp, header)

	uid, err := verifyServiceAuthValue(header, vecSecret)
	if err != nil {
		t.Fatalf("shared-vector header rejected: %v", err)
	}
	if uid != vecUserID {
		t.Fatalf("user_id mismatch: got %q want %q", uid, vecUserID)
	}
}

func TestServiceAuthRejects(t *testing.T) {
	good := mintServiceAuth(vecUserID, vecExp, vecSecret)

	// tampered signature: same shape, different (valid-b64) sig segment
	tParts := strings.Split(good, ".")
	tParts[3] = base64.RawURLEncoding.EncodeToString([]byte("not-the-real-hmac-32-bytes-xxxxx"))
	tamperedSig := strings.Join(tParts, ".")

	// user_id swap: replace the user_id segment, KEEP the original signature
	// (the forgery the binding must defeat)
	sParts := strings.Split(good, ".")
	sParts[1] = base64.RawURLEncoding.EncodeToString([]byte("victim-other-user-id"))
	swappedUID := strings.Join(sParts, ".")

	cases := []struct{ name, raw, secret string }{
		{"wrong-secret", good, "a-totally-different-secret"},
		{"empty-secret-disabled", good, ""},
		{"literal-disabled", good, "disabled"},
		{"tampered-signature", tamperedSig, vecSecret},
		{"user-id-swap-keeps-sig", swappedUID, vecSecret},
		{"expired", mintServiceAuth(vecUserID, time.Now().Unix()-10, vecSecret), vecSecret},
		{"too-few-parts", "v1.abc.123", vecSecret},
		{"bad-version", "v2." + strings.Join(strings.Split(good, ".")[1:], "."), vecSecret},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			if _, err := verifyServiceAuthValue(c.raw, c.secret); !errors.Is(err, ErrServiceAuthInvalid) {
				t.Fatalf("expected ErrServiceAuthInvalid, got %v", err)
			}
		})
	}
}

func TestServiceAuthRoundTripVariousUsers(t *testing.T) {
	exp := time.Now().Unix() + 60
	for _, uid := range []string{
		"2d2dfffdad62ff927abba1295c73a4eab7666813280ea8b356da845e440c41ff",
		"f98d6328e961812226658fa0fb4252d2cd4bb651d3df573c8b48671587b38e2c",
		"x", // short id still binds
	} {
		got, err := verifyServiceAuthValue(mintServiceAuth(uid, exp, vecSecret), vecSecret)
		if err != nil || got != uid {
			t.Fatalf("round-trip failed for %q: got=%q err=%v", uid, got, err)
		}
	}
}
